diff --git a/node_modules/react-pin-field/dist/react-pin-field.cjs.development.js b/node_modules/react-pin-field/dist/react-pin-field.cjs.development.js
index ff37b35..c8766ea 100644
--- a/node_modules/react-pin-field/dist/react-pin-field.cjs.development.js
+++ b/node_modules/react-pin-field/dist/react-pin-field.cjs.development.js
@@ -106,10 +106,12 @@ var ReactPinField = /*#__PURE__*/react.forwardRef(function (customProps, fwdRef)
       ref.current.addEventListener("reject-key", handleRejectKey);
       ref.current.addEventListener("complete", handleComplete);
       return function () {
-        ref.current.removeEventListener("change", handleChange);
-        ref.current.removeEventListener("resolve-key", handleResolveKey);
-        ref.current.removeEventListener("reject-key", handleRejectKey);
-        ref.current.removeEventListener("complete", handleComplete);
+		if (ref.current) {
+			ref.current.removeEventListener("change", handleChange);
+			ref.current.removeEventListener("resolve-key", handleResolveKey);
+			ref.current.removeEventListener("reject-key", handleRejectKey);
+			ref.current.removeEventListener("complete", handleComplete);
+		}
       };
     }
 
diff --git a/node_modules/react-pin-field/dist/react-pin-field.esm.js b/node_modules/react-pin-field/dist/react-pin-field.esm.js
index f74c7d4..8e005f3 100644
--- a/node_modules/react-pin-field/dist/react-pin-field.esm.js
+++ b/node_modules/react-pin-field/dist/react-pin-field.esm.js
@@ -103,10 +103,12 @@ var ReactPinField = /*#__PURE__*/forwardRef(function (customProps, fwdRef) {
       ref.current.addEventListener("reject-key", handleRejectKey);
       ref.current.addEventListener("complete", handleComplete);
       return function () {
-        ref.current.removeEventListener("change", handleChange);
-        ref.current.removeEventListener("resolve-key", handleResolveKey);
-        ref.current.removeEventListener("reject-key", handleRejectKey);
-        ref.current.removeEventListener("complete", handleComplete);
+		if (ref.current) {
+			ref.current.removeEventListener("change", handleChange);
+			ref.current.removeEventListener("resolve-key", handleResolveKey);
+			ref.current.removeEventListener("reject-key", handleRejectKey);
+			ref.current.removeEventListener("complete", handleComplete);
+		}
       };
     }
 
